src/util/parallel/data.mli

```diff
@@ -0,0 +1,32 @@
+open Batteries
+
+(** A type with a default factory *) 
+module type DefaultType = sig
+  type t
+  val default: unit -> t
+  val to_string: t -> string
+end
+
+(** A lock free concurrency safe hashmap *)
```

Comment:
Me with my usual line of thinking: what about [saturn's `Htbl`](https://ocaml-multicore.github.io/saturn/saturn/Saturn/Htbl/index.html)?

Reply:
The restrictions we have (always a default value, never remove things etc) makes this a bit faster. I have an implementation based on the saturn hashmap that I will also include after the solver itself so that I can measure it once more when everything else is final.